File: backend/app/api/v1/endpoints/sessions.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, HTTPException, Request, status

from app.core.realtime_security import create_audio_ingest_token
from app.schemas.realtime import (
    AudioFormat,
    IngestPolicy,
    SessionCreateRequest,
    SessionCreateResponse,
    SourceRegisterResponse,
)
from app.services.realtime_session_store import ExpectedAudio, RealtimeSessionConfig, session_store


router = APIRouter()


def _ws_base_url(request: Request) -> str:
    forwarded_proto = request.headers.get("x-forwarded-proto")
    scheme = (forwarded_proto or request.url.scheme or "http").lower()
    ws_scheme = "wss" if scheme == "https" else "ws"
    host = request.headers.get("host") or request.url.netloc
    return f"{ws_scheme}://{host}".rstrip("/")


@router.post("", response_model=SessionCreateResponse)
async def create_session(payload: SessionCreateRequest, request: Request) -> SessionCreateResponse:
    config = RealtimeSessionConfig(
        language_code=payload.language_code,
        expected_audio=ExpectedAudio(
            codec=payload.audio_encoding,
            sample_rate_hz=payload.target_sample_rate_hz,
            channels=payload.channels,
        ),
        interim_results=payload.interim_results,
        enable_word_time_offsets=payload.enable_word_time_offsets,
    )
    session = session_store.create_with_id(payload.session_id, config) if payload.session_id else session_store.create(config)

    ws_base = _ws_base_url(request)
    ingest_policy = IngestPolicy(
        expected_audio=AudioFormat(
            codec=config.expected_audio.codec,  # type: ignore[arg-type]
            sample_rate_hz=config.expected_audio.sample_rate_hz,
            channels=config.expected_audio.channels,
        ),
        recommended_frame_ms=config.recommended_frame_ms,
        max_frame_ms=config.max_frame_ms,
    )

    return SessionCreateResponse(
        session_id=session.session_id,
        audio_ws_url=f"{ws_base}/api/v1/ws/audio/{session.session_id}",
        frontend_ws_url=f"{ws_base}/api/v1/ws/frontend/{session.session_id}",
        transcript_test_ws_url=f"{ws_base}/api/v1/ws/in-meeting/{session.session_id}",
        ingest_policy=ingest_policy,
    )


@router.post("/{session_id}/sources", response_model=SourceRegisterResponse)
async def register_source(session_id: str, platform: str | None = None) -> SourceRegisterResponse:
    session = session_store.get(session_id)
    if not session:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Session not found")

    token_ttl_seconds = 1800
    token = create_audio_ingest_token(session_id=session_id, ttl_seconds=token_ttl_seconds)
    return SourceRegisterResponse(
        session_id=session_id,
        source_id=str(uuid.uuid4()),
        platform=platform.strip() if platform else None,
        audio_ingest_token=token,
        token_ttl_seconds=token_ttl_seconds,
    )

File: electron/src/renderer/lib/api/sessions.ts
import api from '../apiClient';

export interface SessionCreateRequest {
  session_id?: string;
  language_code?: string;
  target_sample_rate_hz?: number;
  audio_encoding?: 'PCM_S16LE';
  channels?: number;
  realtime?: boolean;
  interim_results?: boolean;
  enable_word_time_offsets?: boolean;
}

export interface SessionCreateResponse {
  session_id: string;
  audio_ws_url: string;
  frontend_ws_url: string;
  transcript_test_ws_url: string;
  ingest_policy: {
    expected_audio: { codec: 'PCM_S16LE'; sample_rate_hz: number; channels: number };
    recommended_frame_ms: number;
    max_frame_ms: number;
  };
}

export interface SourceRegisterResponse {
  session_id: string;
  audio_ingest_token: string;
  token_ttl_seconds: number;
}

export const sessionsApi = {
  create: (data: SessionCreateRequest) =>
    api.post<SessionCreateResponse>('/sessions', data, { skipAuth: true }),

  registerSource: (sessionId: string, platform?: string) => {
    const query = platform ? `?platform=${encodeURIComponent(platform)}` : '';
    return api.post<SourceRegisterResponse>(`/sessions/${sessionId}/sources${query}`, undefined, { skipAuth: true });
  },
};
